release/apparmor.go

```diff
@@ -97,117 +213,130 @@ var (
+		appArmorLevel = UnusableAppArmor
+		appArmorSummary = fmt.Sprintf("apparmor is enabled but required kernel features are missing: %s",
+			strings.Join(missingKernelFeatures, ", "))
+		return
 	}
-	var missing []string
-	for _, feature := range requiredAppArmorFeatures {
-		if !isDirectory(filepath.Join(appArmorFeaturesSysPath, feature)) {
-			missing = append(missing, feature)
+
+	// Next check that the parser supports preferred parser features.
+	// If we have any missing preferred features then apparmor is partially enabled.
+	for _, feature := range preferredAppArmorParserFeatures {
+		if !strutil.SortedListContains(parserFeatures, feature) {
+			missingParserFeatures = append(missingParserFeatures, feature)
+		}
+	}
+	if len(missingParserFeatures) > 0 {
+		appArmorLevel = PartialAppArmor
+		appArmorSummary = fmt.Sprintf("apparmor_parser is available but some features are missing: %s",
+			strings.Join(missingParserFeatures, ", "))
+		return
+	}
+
+	// Lastly check that the kernel supports preferred kernel features.
+	for _, feature := range preferredAppArmorKernelFeatures {
+		if !strutil.SortedListContains(kernelFeatures, feature) {
+			missingKernelFeatures = append(missingKernelFeatures, feature)
 		}
 	}
-	if len(missing) > 0 {
-		return PartialAppArmor, fmt.Sprintf("apparmor is enabled but some features are missing: %s", strings.Join(missing, ", "))
+	if len(missingKernelFeatures) > 0 {
+		appArmorLevel = PartialAppArmor
+		appArmorSummary = fmt.Sprintf("apparmor is enabled but some kernel features are missing: %s",
+			strings.Join(missingKernelFeatures, ", "))
+		return
 	}
-	return FullAppArmor, "apparmor is enabled and all features are available"
+
+	// If we got here then all features are available and supported.
+	appArmorLevel = FullAppArmor
+	appArmorSummary = "apparmor is enabled and all features are available"
 }
 
-// AppArmorFeatures returns a sorted list of apparmor features like
-// []string{"dbus", "network"}.
-func AppArmorFeatures() []string {
+func probeAppArmorKernelFeatures() []string {
 	// note that ioutil.ReadDir() is already sorted
 	dentries, err := ioutil.ReadDir(appArmorFeaturesSysPath)
 	if err != nil {
-		return nil
+		return []string{}
 	}
-	appArmorFeatures := make([]string, 0, len(dentries))
-	for _, f := range dentries {
-		if isDirectory(filepath.Join(appArmorFeaturesSysPath, f.Name())) {
-			appArmorFeatures = append(appArmorFeatures, f.Name())
+	features := make([]string, 0, len(dentries))
+	for _, fi := range dentries {
+		if fi.IsDir() {
+			features = append(features, fi.Name())
 		}
 	}
-	return appArmorFeatures
-}
-
-// parser probe related code
-type apparmorParserFeature struct {
-	feature string
-	rule    string
+	return features
 }
 
-var (
-	requestedParserFeatures = []apparmorParserFeature{
-		{"unsafe", "change_profile unsafe /**,"},
+func probeAppArmorParserFeatures() []string {
+	parser := findAppArmorParser()
+	if parser == "" {
+		return []string{}
```

Comment:
This is the thing that still bothers me. We are returning an empty list to denote that either we can't find a parser but also a parser that was found but doesn't support any special features. With the first, that is always at best UnusableAppArmor while the second could be other things. As it stands, your code elsewhere has "unsafe" as required so this always ends up as UnsableAppArmor, but I'm uncomfortable with this at an API level.

Reply:
I can refactor this to have return an explicit error if you'd prefer that.

Reply:
And to be clear, this can be an internal error that is never surfaced (`errNotFound` or something like that).

Reply:
IME, it can return an error, like it did before, but the caller shouldn't carry that error forward. The caller should make that UnusableAppArmor. With a change along these lines, I'll +1 the PR.

Reply:
Thanks, doing that now.

Reply:
Done